Fix all reported clippy warnings

dnaka91 opened the pull request:
Currently clippy reports many warnings. This PR fixes all of them with the exception of a  
few being allowed now as it bring any benefit to adjust to them.

While doing so I replaced many usages of `PathBuf` with `&Path` as well.

calixteman commented on src/reader.rs:
```diff
@@ -361,12 +361,11 @@ impl Debug for GCNO {
+        Gcno {
             version: 0,
             checksum: 0,
-            cwd: None,
```
I'm not sure that's a good idea to remove this property even if it's unused.

dnaka91 replied:
I stumbled over this at first and quickly got some test failures.

Everything is good as long as the value for still read [here](https://github.com/mozilla/grcov/pull/625/files#diff-267c9f4da66412a9f439ac08d224356fe24265b5e1cebb6c44c2d55b96414513R454). I simply removed it from the struct but the reading/parsing logic _should_ not be affected.

Of course I can put it back in if you prefer.

calixteman replied:
Yep I saw that the value is still read.
It'd prefer keeping it in case we need it or even for debug purposes.

dnaka91 replied:
I put it back in now. Interestingly clippy didn't warn about it anymore now 🤔

calixteman commented on src/output.rs:
```diff
@@ -96,7 +96,7 @@ pub fn output_activedata_etl(results: CovResultIter, output_file: Option<&str>,
             start_indexes.push(function.start);
         }
-        start_indexes.sort();
+        start_indexes.sort_unstable();
```
I suppose that all `function.start` are different so I guess that using `sort_unstable` doesn't make any difference in the final order with `sort`. Out of curiosity, do you have a link on some benchmarks to compare perf for an array with unique elements ?
Anyway we should maybe use a `btreeset` here.

dnaka91 replied:
I don't have any benchmarks for that, mostly just followed clippy's recommendations.
It probably doesn't make a difference in this case and just satisfies Google 😅

calixteman commented on src/parser.rs:
```diff
@@ -529,10 +531,15 @@ fn get_xml_attribute<R: BufRead>(
     )))
 }
 
+struct JacocoReport {
```
nit: we use to have this kind of stuff in defs.rs.

dnaka91 replied:
I moved it over to defs.rs 🙇

calixteman commented on src/parser.rs:
```diff
@@ -865,11 +873,11 @@ mod tests {
         let func = result.functions.get("MainProcessSingleton").unwrap();
         assert_eq!(func.start, 15);
-        assert_eq!(func.executed, true);
+        assert!(func.executed);
```
We'll lose the benefit of using `pretty_assertions` here, but do we care ?? @marco-c what's your opinion ?

marco-c replied:
I guess they're very few cases, so we can live with it if it's what clippy wants

dnaka91 replied:
Alternatively I can simply turn of that specific clippy lint if that's more preferable.

calixteman reviewed (approved):
All in all, looks good for me with all tests green.
Thanks a lot for doing that.

dnaka91 commented:
I have updated the PR to fix the windows build errors and added a few fixes for lints coming from the nightly version of clippy.
Anything else to do for this to be merged @calixteman?

I have a few more changes piled up and based on this PR. Would be lovely if this could be merged soonish 🙇‍♂️

marco-c commented:
@dnaka91 I think we're good, could you rebase on top of current master?

dnaka91 commented:
@marco-c done, just rebased on top of master 🙇
